daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/FNFunctions.scala

```diff
@@ -683,26 +683,109 @@ case class FNLocalName1(recipe: CompiledDPath, argType: NodeInfo.Kind)
   }
 
   override def run(dstate: DState) {
-    // Save off original state, which is the original
-    // element/node that calls inputValueCalc with fn:local-name
-    //
-    val origState = dstate
+    // Save off original node, which is the original
+    // element/node that calls fn:local-name
+    val savedNode = dstate.currentNode
 
     // Execute the recipe/expression which should
-    // return a node/element whose local-name we want.
-    //
+    // return a node/element whose local-name we want
     recipe.run(dstate)
 
     val localName = dstate.currentElement.name
+    dstate.setCurrentNode(savedNode)
 
     if (localName.contains(":"))
       throw new IllegalArgumentException("fn:local-name failed. " + localName + " is not a valid NCName as it contains ':'.")
 
-    // The original state contains the node/element upon which
-    // fn:local-name was called.  This is where we should set
-    // the value.
-    //
-    origState.setCurrentValue(localName)
+    dstate.setCurrentValue(localName)
+  }
+}
+
+/**
+ * Returns the namespace URI of the name of \$arg as an xs:string
+ * value.
+ *
+ * If the argument is omitted, it defaults to the context item (.).
+ * The behavior of the function if the argument is omitted is
+ * exactly the same as if the context item had been passed as
+ * the argument.
+ *
+ * If the node identified by \$arg is neither an element nor an
+ * attribute node, or it is an element or attribute node whose
+ * expanded-QName is in no namespace, then the function returns
+ * the zero-length xs:string value.
+ *
+ * Otherwise, the result will be the namespace URI of the
+ * expanded-QName of the node identified by \$arg returned as an
+ * xs:string value.
+ *
+ * The following errors may be raised when \$arg is omitted:
+ * - If the context item is absent, dynamic error [err:XPDY002]
+ * - If the context item is not a node, type error [err:XPTY004]
```

Comment:
This comment seems to contradict the statement above that says if the node identified by arg is neither an element nor an attribute node... zero-length string is the value. This appears to be the implemented behavior as evidenced by your test that uses element "e5" which has fn:namespace-uri(0) which by passing an int here, makes no sense whatsoever, but you've said it returns empty string in that case, which it does.

But 0 is not a node, so this second comment suggests this should be a type error. Personally, that sounds right to me. If you pass any value to this thing you should get a type error. I think. The argument should either be absent, or be a path to a node that exists.

Reply:
Yes, I agree with you about the type error XPTY004.  I was just looking through the DFDL spec this morning and I saw the following paragraph in 23. Expression Language:

> DFDL implementations MUST comply with the error code behaviour in Appendix G of the XPath 2.0 spec and map these to the correct DFDL failure type. All but one of XPath's errors map to a schema definition error. The exception is XPTY0004, which is used both for static and dynamic cases of type mismatch. A static type mismatch maps to a schema definition error, whereas a dynamic type mismatch maps to a processing error. A DFDL implementation should distinguish the two kinds of XPTY0004 error if it is able to do so, but if unable it should map all XPTY0004 errors to a schema definition error.

I'll issue a dstate.SDE if dstate.currentNode is null, saying something like (please suggest any better or more Daffodil-style way to say it):

fn:namespace-uri [ context item | argument ] '%s' is not a node

Reply:
Suggest use the term "element". The only kinds of nodes in DFDL are elements. The error in this case is passing a value, not a path to an element. So the message can say "is not an element."

Reply:
One other thing worth pointing out, I *think* there is a current limiation in our DPath implementation where we can't tell if a subexpression results to a node, we only keep track of the primitive type. For example, we can't tell the difference between
```
fn:namespace-uri(0)
```
and
```
fn:namespace-uri(/path/to/int/element)
```
In both cases, I think all our implementation knows at schema compile time is that the result of the argument is an int, we can't tell that one was a constant int and the other is an element with an int type. And you can't just check if the subexpression is constant, because the expression could be something like this:
```
fn:namespace-uri(
   if (somecondition) then /path/to/int/element
   else /path/to/string/element)
)
```
I wonder if to support this check statically, our NodeInfo type system needs a distinction between node types and non-node types, e.g. NodeInfo.Int vs NodeInfo.NodeInt. That sounds like a big change so a runtime check is very reasonable for this change.

Reply:
Agree. Runtime check is fine for this. The NodeType.Exists was supposed to be able to make this compilation-time distinction, .... but the compiled expression result type is probably not preserving enough about whether the expression is only a value (e.g.,, output of a + b is only a value), or is a path that could be evaluated for a node or a value e.g., for fn:exists, fn:namespace-uri, etc.